eth/filters/api.go

```diff
@@ -280,8 +328,225 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
+				}
+				notifyLogsIf(notifier, rpcSub.ID, logs, &hashes)
+				// Assuming batch = all logs of a single block
+				delivered = logs[0].BlockNumber
+
+			case <-rpcSub.Err(): // client send an unsubscribe request
+				return
+			case <-notifier.Closed(): // connection dropped
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// doHistLogs retrieves the logs older than current header, and forward them to the histLogs channel.
+func (api *FilterAPI) doHistLogs(ctx context.Context, from int64, addrs []common.Address, topics [][]common.Hash, histLogs chan<- []*types.Log) error {
+	// Fetch logs from a range of blocks.
+	fetchRange := func(from, to int64) error {
+		f := api.sys.NewRangeFilter(from, to, addrs, topics)
+		logsCh, errChan := f.rangeLogsAsync(ctx)
+		for {
+			select {
+			case logs := <-logsCh:
+				select {
+				case histLogs <- logs:
+				case <-ctx.Done():
+					// Flush out all logs until the range filter voluntarily exits.
+					continue
+				}
+			case err := <-errChan:
+				return err
+			}
+		}
+	}
+
+	for {
+		// Get the latest block header.
+		header := api.sys.backend.CurrentHeader()
+		if header == nil {
+			return errors.New("unexpected error: no header block found")
+		}
+		head := header.Number.Int64()
+		if from > head {
+			logger.Info("Finish historical sync", "from", from, "head", head)
+			return nil
+		}
+		if err := fetchRange(from, head); err != nil {
+			if errors.Is(err, context.Canceled) {
+				logger.Info("Historical logs delivery canceled", "from", from, "to", head)
+				return nil
+			}
+			return err
+		}
+		// Move forward to the next batch
+		from = head + 1
+	}
+}
+
+// notifyLogsIf sends logs to the notifier if the condition is met.
+// It assumes all logs of the same block are either all removed or all added.
+func notifyLogsIf(notifier notifier, id rpc.ID, logs []*types.Log, hashes *lru.BasicLRU[common.Hash, struct{}]) {
+	// Iterate logs and batch them by block hash.
+	type batch struct {
+		start   int
+		end     int
+		hash    common.Hash
+		removed bool
+	}
+	var (
+		batches = make([]batch, 0)
+		h       common.Hash
+	)
+	for i, log := range logs {
+		if h == log.BlockHash {
+			// Skip logs of seen block
+			continue
+		}
+		if len(batches) > 0 {
+			batches[len(batches)-1].end = i
+		}
+		batches = append(batches, batch{start: i, hash: log.BlockHash, removed: log.Removed})
+		h = log.BlockHash
+	}
+	// Close off last batch.
+	if batches[len(batches)-1].end == 0 {
+		batches[len(batches)-1].end = len(logs)
+	}
+	for _, batch := range batches {
+		if hashes != nil {
+			// During reorgs it's possible that logs from the new chain have been delivered.
+			// Avoid sending removed logs from the old chain and duplicate logs from new chain.
+			if batch.removed && !hashes.Contains(batch.hash) {
+				continue
+			}
+			if !batch.removed && hashes.Contains(batch.hash) {
+				continue
+			}
+			hashes.Add(batch.hash, struct{}{})
+		}
+		for _, log := range logs[batch.start:batch.end] {
+			log := log
```

Comment:
Do we need to deref the log here to copy it? 
e.g. `log := *log`
Otherwise we just copy the pointer if I understand correctly